### PR TITLE
Engine

### PR DESCRIPTION
**Pull request by daria-sieroshtan**

### CODE REVIEW

File: src/Snowflakes/Engine.php [hunk 1]
```diff
@@ -4,30 +4,162 @@
...
     *
     * @param int            $width
     * @param int            $height
     * @param int            $numMolecules If not provided the reasonable default is used
      *@param array|null    $crystalCenters
     */
     public function __construct($width, $height, $numMolecules = null, $crystalCenters = null) {
-        throw new \RuntimeException("Not implemented");
+        if ( ! is_int($width) || $width <= 0) {
+            throw new \RuntimeException("Width should be positive integer");
```
Comment (arodiss) [inline src/Snowflakes/Engine.php]: @daria-sieroshtan 
PHP has [special exception class](https://www.php.net/manual/en/class.invalidargumentexception.php)

File: src/Snowflakes/Engine.php [hunk 2]
```diff
@@ -4,30 +4,162 @@
...
+        $moleculesYs = range(0, $height);
+
+        $generatedMolecules = [];
+        foreach ($moleculesXs as $x) {
+            foreach ($moleculesYs as $y) {
+                $generatedMolecules[] = new Molecule($x, $y);
+            }
+        }
+
+        shuffle($generatedMolecules);
```
Comment (arodiss) [inline src/Snowflakes/Engine.php:82]: They are already random, why shuffle?

File: src/Snowflakes/Molecule.php [hunk 3]
```diff
@@ -6,25 +6,77 @@ class Molecule
 {
     const MOLECULE_STATE_FREE = 0;
     const MOLECULE_STATE_FROZEN = 1;
+    const MOLECULE_STATE_CRYSTAL_CENTER = 2;
+
+    private $x;
+    private $y;
+    private $state = 0;
```
Comment (arodiss) [inline src/Snowflakes/Molecule.php]: @daria-sieroshtan 
magic var

**Review by arodiss — APPROVED** [review]

File: src/Snowflakes/Engine.php (hunk 2 above)
Comment (daria-sieroshtan) [reply src/Snowflakes/Engine.php:82]: They are not random initially.

**Pull request merged**